File: reference-python/cannon/viz.py
"""Mermaid emitters — the visualization surface.

No UI server, no canvas: Mermaid-in-Markdown renders in GitHub, VS Code, Obsidian
and most doc viewers, and the underlying JSON is there for a real UI later. This
is the lightweight, version-controllable answer to "visualizing threat models
and attack chains."
"""
from __future__ import annotations

import re

from .artifacts import Chain, TriagedFinding


def _node_id(name: str, prefix: str = "n") -> str:
    h = re.sub(r"[^a-zA-Z0-9]+", "_", name).strip("_") or "x"
    return f"{prefix}_{h[:32]}"


def _esc(s: str) -> str:
    return (s or "").replace('"', "'").replace("\n", " ").strip()


def threat_model_mermaid(components, flows, boundaries) -> str:
    """flowchart of components + data flows, grouped by trust where given."""
    if not components and not flows:
        return ""
    lines = ["```mermaid", "flowchart LR"]

    # Group components into trust subgraphs when trust labels exist.
    by_trust: dict[str, list] = {}
    for c in components:
        by_trust.setdefault(c.trust or "system", []).append(c)

    declared: set[str] = set()
    for trust, comps in by_trust.items():
        if trust and trust != "system":
            lines.append(f'  subgraph {_node_id(trust, "tb")}["{_esc(trust)}"]')
            indent = "    "
        else:
            indent = "  "
        for c in comps:
            nid = _node_id(c.name)
            declared.add(c.name)
            lines.append(f'{indent}{nid}["{_esc(c.name)}"]')
        if trust and trust != "system":
            lines.append("  end")

    for fl in flows:
        a, b = _node_id(fl.src), _node_id(fl.dst)
        # Declare any endpoint not already a known component.
        if fl.src not in declared:
            lines.append(f'  {a}["{_esc(fl.src)}"]'); declared.add(fl.src)
        if fl.dst not in declared:
            lines.append(f'  {b}["{_esc(fl.dst)}"]'); declared.add(fl.dst)
        if fl.label:
            lines.append(f'  {a} -->|"{_esc(fl.label)}"| {b}')
        else:
            lines.append(f"  {a} --> {b}")

    lines.append("```")
    return "\n".join(lines)


def chains_mermaid(chains: list[Chain]) -> str:
    """One left-to-right path per chain: premise → step → … → impact."""
    if not chains:
        return ""
    lines = ["```mermaid", "flowchart LR"]
    for ci, c in enumerate(chains):
        sg = _node_id(c.title, f"c{ci}")
        lines.append(f'  subgraph {sg}["{_esc(c.title)} ({c.severity})"]')
        prev = f"{sg}_start"
        lines.append(f'    {prev}(["{_esc(c.premise or "attacker start")[:60]}"])')
        for si, step in enumerate(c.steps):
            nid = f"{sg}_s{si}"
            label = step.action or step.title or "step"
            lines.append(f'    {nid}["{_esc(label)[:70]}"]')
            lines.append(f"    {prev} --> {nid}")
            prev = nid
        end = f"{sg}_impact"
        lines.append(f'    {end}(["💥 {_esc(c.impact or "impact")[:60]}"])')
        lines.append(f"    {prev} --> {end}")
        lines.append("  end")
    lines.append("```")
    return "\n".join(lines)


def triage_table(triaged: list[TriagedFinding]) -> str:
    """Markdown table of triaged findings, ranked."""
    if not triaged:
        return "_No findings._"
    rows = ["| Rank | Sev | Verdict | Conf | Corrob | Finding | Location |",
            "|---:|---|---|---:|---:|---|---|"]
    for i, t in enumerate(triaged, 1):
        f = t.accumulated.representative
        loc = f.file + (f":{f.line}" if f.line else "")
        mark = "✅" if t.confirmed else ("❌" if t.verdict.verdict == "FALSE_POSITIVE" else "❔")
        rows.append(
            f"| {i} | {t.accumulated.max_severity} | {mark} {t.verdict.verdict} | "
            f"{t.verdict.confidence:.2f} | {t.accumulated.corroboration} | "
            f"{_esc(f.title)} | `{_esc(loc)}` |"
        )
    return "\n".join(rows)

File: src/viz.rs
//! Mermaid emitters for REPORT.md (port of viz.py). The TUI draws its own
//! native graphs (see tui/graph.rs); this is the markdown/HTML surface.

use crate::artifacts::{Chain, Component, DataFlow, TriagedFinding};
use std::collections::BTreeMap;

fn node_id(name: &str, prefix: &str) -> String {
    let mut h = String::new();
    let mut prev = false;
    for c in name.chars() {
        if c.is_ascii_alphanumeric() {
            h.push(c);
            prev = false;
        } else if !prev {
            h.push('_');
            prev = true;
        }
    }
    let h = h.trim_matches('_');
    let h: String = h.chars().take(32).collect();
    format!("{prefix}_{}", if h.is_empty() { "x".to_string() } else { h })
}

fn esc(s: &str) -> String {
    s.replace('"', "'").replace('\n', " ").trim().to_string()
}

pub fn threat_model_mermaid(components: &[Component], flows: &[DataFlow], boundaries: &[String]) -> String {
    let _ = boundaries;
    if components.is_empty() && flows.is_empty() {
        return String::new();
    }
    let mut lines = vec!["```mermaid".to_string(), "flowchart LR".to_string()];

    let mut by_trust: BTreeMap<String, Vec<&Component>> = BTreeMap::new();
    for c in components {
        let key = if c.trust.is_empty() { "system".to_string() } else { c.trust.clone() };
        by_trust.entry(key).or_default().push(c);
    }

    let mut declared: Vec<String> = Vec::new();
    for (trust, comps) in &by_trust {
        let grouped = trust != "system";
        let indent = if grouped { "    " } else { "  " };
        if grouped {
            lines.push(format!("  subgraph {}[\"{}\"]", node_id(trust, "tb"), esc(trust)));
        }
        for c in comps {
            lines.push(format!("{indent}{}[\"{}\"]", node_id(&c.name, "n"), esc(&c.name)));
            declared.push(c.name.clone());
        }
        if grouped {
            lines.push("  end".to_string());
        }
    }

    for fl in flows {
        let a = node_id(&fl.src, "n");
        let b = node_id(&fl.dst, "n");
        if !declared.contains(&fl.src) {
            lines.push(format!("  {a}[\"{}\"]", esc(&fl.src)));
            declared.push(fl.src.clone());
        }
        if !declared.contains(&fl.dst) {
            lines.push(format!("  {b}[\"{}\"]", esc(&fl.dst)));
            declared.push(fl.dst.clone());
        }
        if fl.label.is_empty() {
            lines.push(format!("  {a} --> {b}"));
        } else {
            lines.push(format!("  {a} -->|\"{}\"| {b}", esc(&fl.label)));
        }
    }
    lines.push("```".to_string());
    lines.join("\n")
}

pub fn chains_mermaid(chains: &[Chain]) -> String {
    if chains.is_empty() {
        return String::new();
    }
    let mut lines = vec!["```mermaid".to_string(), "flowchart LR".to_string()];
    for (ci, c) in chains.iter().enumerate() {
        let sg = node_id(&c.title, &format!("c{ci}"));
        lines.push(format!("  subgraph {sg}[\"{} ({})\"]", esc(&c.title), c.severity));
        let mut prev = format!("{sg}_start");
        let premise: String = esc(if c.premise.is_empty() { "attacker start" } else { &c.premise }).chars().take(60).collect();
        lines.push(format!("    {prev}([\"{premise}\"])"));
        for (si, step) in c.steps.iter().enumerate() {
            let nid = format!("{sg}_s{si}");
            let label_src = if !step.action.is_empty() { &step.action } else { &step.title };
            let label: String = esc(label_src).chars().take(70).collect();
            lines.push(format!("    {nid}[\"{label}\"]"));
            lines.push(format!("    {prev} --> {nid}"));
            prev = nid;
        }
        let end = format!("{sg}_impact");
        let impact: String = esc(if c.impact.is_empty() { "impact" } else { &c.impact }).chars().take(60).collect();
        lines.push(format!("    {end}([\"💥 {impact}\"])"));
        lines.push(format!("    {prev} --> {end}"));
        lines.push("  end".to_string());
    }
    lines.push("```".to_string());
    lines.join("\n")
}

pub fn triage_table(triaged: &[TriagedFinding]) -> String {
    if triaged.is_empty() {
        return "_No findings._".to_string();
    }
    let mut rows = vec![
        "| Rank | Sev | Verdict | Conf | Corrob | Finding | Location |".to_string(),
        "|---:|---|---|---:|---:|---|---|".to_string(),
    ];
    for (i, t) in triaged.iter().enumerate() {
        let f = &t.accumulated.representative;
        let mark = if t.confirmed() {
            "✅"
        } else if t.verdict.verdict == "FALSE_POSITIVE" {
            "❌"
        } else {
            "❔"
        };
        rows.push(format!(
            "| {} | {} | {} {} | {:.2} | {} | {} | `{}` |",
            i + 1,
            t.accumulated.max_severity,
            mark,
            t.verdict.verdict,
            t.verdict.confidence,
            t.accumulated.corroboration,
            esc(&f.title),
            esc(&f.loc())
        ));
    }
    rows.join("\n")
}
